src/shave.cpp

```diff
@@ -354,9 +358,22 @@ void filterFeatures(vtzero::tile_builder* finalvt,
             return true; // skip to next feature
         }
 
-        // If evaluate() returns true, this feature includes properties that are relevant to the filter.
-        // So we add the feature to the final layer.
-        if (evaluate(mbgl_filter_obj, zoom, geometry_type, feature)) {
+        const mbgl::style::expression::EvaluationResult result = evaluate(mbgl_filter_obj, zoom, geometry_type, feature);
+        // expression hit error
+        if (!result) {
+            std::ostringstream msg;
+            msg << "[vtshaver] filter error: '" << result.error().message << "' skipping feature";
+            if (feature.has_id()) {
+              msg << " " << feature.id();
+            }
+            msg << " in " << std::string{layer.name()};
+            std::clog << msg.str() << "\n";
+            return true;
+        }
+
+        const mbgl::optional<bool> typed = mbgl::style::expression::fromExpressionValue<bool>(*result);
+        bool match = typed ? *typed : false;
```

Comment:
This fancy type logic is taken/matches https://github.com/mapbox/mapbox-gl-native/blob/647959a85327660a8d0b35da0b8a247820a36c15/src/mbgl/style/filter.cpp#L14.